File: app/src/main/java/piotr/michalkiewicz/mealplannerclient/view/main_menu/fragments/CookbookScreenFragment.java
package piotr.michalkiewicz.mealplannerclient.view.main_menu.fragments;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import piotr.michalkiewicz.mealplannerclient.R;
import piotr.michalkiewicz.mealplannerclient.recipes.model.MealTimeRecipe;
import piotr.michalkiewicz.mealplannerclient.recipes.service_generator.RecipeServiceGenerator;
import piotr.michalkiewicz.mealplannerclient.view.main_menu.RecipesRecyclerViewScrollListener;
import piotr.michalkiewicz.mealplannerclient.view.recipes.RecipeActivity;
import piotr.michalkiewicz.mealplannerclient.view.recipes.adapters.RecipeRecyclerViewAdapter;
import piotr.michalkiewicz.mealplannerclient.view.utils.InitializableViewWithCategory;
import piotr.michalkiewicz.mealplannerclient.view.main_menu.presenters.CookbookFragmentPresenter;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class CookbookScreenFragment extends Fragment implements InitializableViewWithCategory<List<MealTimeRecipe>> {

    private ViewGroup recipesLayoutContainer;
    private CookbookFragmentPresenter presenter;
    private final int itemViewCacheSize = 20;
    private final RecipeServiceGenerator recipeServiceGenerator = new RecipeServiceGenerator(getContext());

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_cookbook_screen, container, false);
        init(view);

        return view;
    }

    @Override
    public void onDestroy(){
        super.onDestroy();
    }

    private void init(View root){
        assignUIElements(root);
        presenter = new CookbookFragmentPresenter(this);
        presenter.initWithDefaultCategories();
    }

    private void assignUIElements(View root){
        recipesLayoutContainer = root.findViewById(R.id.recipesByCategoriesLayoutContainer);
    }

    @Override
    public void initWithData(List<MealTimeRecipe> data, String category) {
        ViewGroup recipesView = createRecipesHorizontalRecyclerView(data);
        recipesView.setOnClickListener(v->{
            Intent intent = new Intent(getActivity(), RecipeActivity.class);
            startActivity(intent);
        });
        recipesLayoutContainer.addView(recipesView);
    }

    private ViewGroup createRecipesHorizontalRecyclerView(List<MealTimeRecipe> data){
        RecyclerView recyclerView = (RecyclerView) LayoutInflater.from(getContext())
                .inflate(R.layout.recipes_list_recycler_view, recipesLayoutContainer, false);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(new RecipeRecyclerViewAdapter(data));
        recyclerView.addOnScrollListener(new RecipesRecyclerViewScrollListener(getContext()){
            @Override
            public void onLoadMore(int pageNumber, @NotNull RecyclerView recyclerView) {
                recipeServiceGenerator.getRecipesForDiet("standard", new Callback<List<MealTimeRecipe>>() {
                    @Override
                    public void onResponse(Call<List<MealTimeRecipe>> call, Response<List<MealTimeRecipe>> response) {
                  //      recyclerView.getAdapter().
                    }

                    @Override
                    public void onFailure(Call<List<MealTimeRecipe>> call, Throwable t) {

                    }
                }, pageNumber);
            }
        });
        recyclerView.setItemViewCacheSize(itemViewCacheSize);
        return recyclerView;
    }


}

File: app/src/main/java/piotr/michalkiewicz/mealplannerclient/view/main_menu/fragments/CookbookScreenFragment.kt
package piotr.michalkiewicz.mealplannerclient.view.main_menu.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.paging.PagingDataAdapter
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.fragment_cookbook_screen.*
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import piotr.michalkiewicz.mealplannerclient.R
import piotr.michalkiewicz.mealplannerclient.databinding.FragmentCookbookScreenBinding
import piotr.michalkiewicz.mealplannerclient.recipes.Injection
import piotr.michalkiewicz.mealplannerclient.recipes.model.MealTimeRecipeMiniatureData
import piotr.michalkiewicz.mealplannerclient.recipes.model.RecipeParamsPair
import piotr.michalkiewicz.mealplannerclient.view.recipes.paging.OnPrependDataLoadedListener
import piotr.michalkiewicz.mealplannerclient.view.recipes.paging.RecipesAdapter
import piotr.michalkiewicz.mealplannerclient.view.recipes.paging.RecipesSearchViewModel
import java.util.*


class CookbookScreenFragment : Fragment() {

    private lateinit var binding: FragmentCookbookScreenBinding
    private lateinit var viewModel: RecipesSearchViewModel

    private var recipeListsInitParams = LinkedList<RecipeParamsPair>()

    private val onPrependDataLoadedListener = object : OnPrependDataLoadedListener {
        override fun onPrependDataLoaded() {
            val nextRecipeListParameters = recipeListsInitParams.poll()
            if (nextRecipeListParameters != null) {
                attachRecipesRecyclerView(nextRecipeListParameters.category, nextRecipeListParameters.value)
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_cookbook_screen, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding = FragmentCookbookScreenBinding.inflate(layoutInflater)

        viewModel = ViewModelProvider(this, Injection.provideViewModelFactory(onPrependDataLoadedListener))
                .get(RecipesSearchViewModel::class.java)

        initRecipeRecyclerViews()
    }

    private fun launchCoroutineByDietForRecyclerView(recyclerView: RecyclerView, queryParam: String) {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.recipesByDietApiData(queryParam).collect {
                it.let {
                    (recyclerView.adapter as PagingDataAdapter<MealTimeRecipeMiniatureData, RecyclerView.ViewHolder>)
                            .submitData(it)
                }
            }
        }
    }

    private fun initListsParamsWithMockValues() {
        recipeListsInitParams.add(RecipeParamsPair("diet", "STANDARD"))
        recipeListsInitParams.add(RecipeParamsPair("diet", "VEGETARIAN"))
        recipeListsInitParams.add(RecipeParamsPair("diet", "PALEO"))
    }

    private fun initRecipeRecyclerViews() {
        initListsParamsWithMockValues()
        attachRecipesRecyclerView("", "")
    }

    private fun launchCoroutineByTypeForRecyclerView(recyclerView: RecyclerView, queryParam: String) {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.recipesByTypeApiData(queryParam).collect {
                it.let {
                    (recyclerView.adapter as PagingDataAdapter<MealTimeRecipeMiniatureData, RecyclerView.ViewHolder>)
                            .submitData(it)
                }
            }
        }
    }

    private fun launchCoroutineByTagForRecyclerView(recyclerView: RecyclerView, queryParam: String) {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.recipesByTagApiData(queryParam).collect {
                it.let {
                    (recyclerView.adapter as PagingDataAdapter<MealTimeRecipeMiniatureData, RecyclerView.ViewHolder>)
                            .submitData(it)
                }
            }
        }
    }

    private fun launchCoroutineAllRecipesForRecyclerView(recyclerView: RecyclerView) {
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.allRecipesApiData().collect {
                it.let {
                    (recyclerView.adapter as PagingDataAdapter<MealTimeRecipeMiniatureData, RecyclerView.ViewHolder>)
                            .submitData(it)
                }
            }
        }
    }

    private fun createHorizontalRecipesList(labelText: String): View {
        val root = layoutInflater.inflate(R.layout.horizontal_recipes_list,
                recipesByCategoriesLayoutContainer, true) as ViewGroup

        val view = root.getChildAt(root.childCount - 1)

        if (labelText.isEmpty()) {
            view.findViewById<TextView>(R.id.recipesRecyclerViewLabel).text = resources.getString(R.string.all)
        } else {
            view.findViewById<TextView>(R.id.recipesRecyclerViewLabel).text = labelText
        }
        val horizontalRecyclerView = view.findViewById<RecyclerView>(R.id.recipesHorizontalRecyclerView)
        horizontalRecyclerView.layoutManager =
                LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        horizontalRecyclerView.adapter = RecipesAdapter()

        return view
    }

    private fun attachRecipesRecyclerView(category: String, categoryValue: String) {
        val recipesHorizontalListWithLabel = createHorizontalRecipesList(categoryValue)

        if (category.isEmpty() || categoryValue.isEmpty()) {
            launchCoroutineAllRecipesForRecyclerView(recipesHorizontalListWithLabel
                    .findViewById(R.id.recipesHorizontalRecyclerView))
        }

        when (category) {
            "diet" -> launchCoroutineByDietForRecyclerView(recipesHorizontalListWithLabel
                    .findViewById(R.id.recipesHorizontalRecyclerView), categoryValue)
            "type" -> launchCoroutineByTypeForRecyclerView(recipesHorizontalListWithLabel
                    .findViewById(R.id.recipesHorizontalRecyclerView), categoryValue)
            "tag" -> launchCoroutineByTagForRecyclerView(recipesHorizontalListWithLabel
                    .findViewById(R.id.recipesHorizontalRecyclerView), categoryValue)
        }
    }
}